main/ogn-aprs.go

```diff
@@ -0,0 +1,240 @@
+				// filter = "#filter?\r\n"
+				fmt.Fprintf(c, filter)
+			}
+		}
+	}()
+}
+
+func aprsListen() {
+	// https://regex101.com/r/Cv9mSq/1
+	// rex := regexp.MustCompile(`(ICA|FLR|SKY|PAW|FNT)([\dA-Z]{6})>[A-Z]+,qAS,([\d\w]+):\/(\d{6})h(\d*\.?\d*[NS])[\/\\](\d*\.?\d*[EW])['\^nX](\d{3})\/(\d{3})\/A=\d*\s!W(\d+)!\sid([\dA-F]{8})`)
+	rex := regexp.MustCompile(
+		`(ICA|FLR|SKY|PAW|OGN|RND|FMT|MTK|XCG|FAN|FNT)([\dA-Z]{6})>` + // protocol, id
+			// `(SKY)([\dA-Z]{6})>` + // protocol, id
+			`[A-Z]+,qAS,([\d\w]+):[\/]` +
+			`(\d{6})h(\d*\.?\d*[NS])[\/\\](\d*\.?\d*[EW])` + // time, lon, lat
+			// `[#&0>AW\^_acnsuvz]` +
+			`\D` + // sep
+			`((\d{3})\/(\d{3})\/A=(\d*))*` + // optional track, speed, alt
+			`(\s!W(\d+)!\s)*` + // optional lon lat precision
+			`(id([\dA-F]{8}))*`) // optional id
+	for {
+		if !globalSettings.APRS_Enabled || OGNDev == nil {
+			// wait until APRS is enabled
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		log.Printf("aprs connecting...")
+		conn, err := net.Dial("tcp", "aprs.glidernet.org:14580")
+		// conn, err := net.Dial("tcp6", "[2a01:4f8:160:6065:229]:14580")
+		if err != nil { // Local connection failed.
+			time.Sleep(3 * time.Second)
+			continue
+		}
+		log.Printf("APRS connected")
+		authenticate(conn)
+		log.Printf("APRS authentication sent...")
+		keepalive(conn)
+		updateFilter(conn)
+
+		aprsReader := bufio.NewReader(conn)
+
+		log.Printf("APRS successfully connected")
+		globalStatus.APRS_connected = true
+
+		// Make sure the exit channel is empty, so we don't exit immediately
+		for len(aprsExitChan) > 0 {
+			<-aprsExitChan
+		}
+
+		go func() {
+			scanner := bufio.NewScanner(aprsReader)
+			for scanner.Scan() {
+				var temp string = scanner.Text()
+				// log.Printf(temp)
+				select {
+				case aprsIncomingMsgChan <- temp: // Put in the channel unless it is full
+				default:
+					fmt.Println("aprsIncomingMsgChan full. Discarding " + temp)
+				}
+			}
+			if scanner.Err() != nil {
+				log.Printf("APRS issue: " + scanner.Err().Error())
+			}
+			aprsExitChan <- true
+		}()
+
+		pgrmzTimer := time.NewTicker(100 * time.Millisecond)
+
+	loop:
+		for globalSettings.APRS_Enabled {
+			// log.Println(len(aprsIncomingMsgChan))
+			select {
+			case data := <-aprsIncomingMsgChan:
+				// APRS,qAS: aircraft beacon
+				// APRS,TCPIP*,qAC: ground station beacon
+
+				res := rex.FindStringSubmatch(data)
+				if res == nil { // no match
+					if strings.Contains(data, "TCPIP*") {
+						// log.Printf("GW data: " + data)
+					} else {
+						log.Printf("No match for: " + data)
+					}
+					continue
+				} else if len(res) == 0 { // no group capture
+					log.Printf("No group capture: " + data)
+				} else if len(res) > 0 && len(res[14]) > 0 {
+					// for i := range res {
+					// 	fmt.Printf("%s(%d)|", res[i], len(res[i]))
+					// }
+					// fmt.Printf("\n")
+					// fmt.Printf("%+v\n", res)
+
+					lat, err := strconv.ParseFloat(res[5][:2], 64)
```

Comment:
Hmm, again, regex named captures could really help readability here - hard to keep track with all these indices